examples/step-57/step-57.cc

```diff
@@ -0,0 +1,948 @@
+    gmres.solve (system_matrix,
+                 newton_update,
+                 system_rhs,
+                 preconditioner);
+
+    std::cout << " ****FGMRES steps: " << solver_control.last_step() << std::endl;
+
+    if (initial_step)
+      {
+        nonzero_constraints.distribute(newton_update);
+      }
+
+    else
+      {
+        zero_constraints.distribute(newton_update);
+      }
+
+  }
+
+  // @sect4{Navier_Stokes_Newton<dim>::newton_iteration}
+  // This part discribe Newton's iteration with given tolerance, bound of iteration and bound of meshes
+  // (if the bound of mesh is two, that means the mesh will be refined once). "initial" is the flag to tell us
+  // whether "setup_system" is necessary, and which part, system matrix or right hand side vector, should be assembled.
+  // If we do line search, the right hand side is already assembled while checking residual norm in last iteration.
+  // Therefore, we just assemble the system matrix at current iteration. If the number of meshes is not one, we must
+  // assemble the whole system at the beginning. And this system depends on present solution which is transfered from
+  // the solution on last mesh. The last argument "output_result" is for output.
+
+  template <int dim>
+  void Navier_Stokes_Newton<dim>::newton_iteration(const double tolerance,
+                                                   const unsigned int max_iteration,
+                                                   const unsigned int n_refinements,
+                                                   bool  is_initial_step,
+                                                   bool  output_result)
+  {
+    double current_res;
+    double last_res;
+    bool   first_step = is_initial_step;
+
+    for (unsigned int refinement = 0; refinement < n_refinements+1; ++refinement)
+      {
+        unsigned int outer_iteration = 0;
+        last_res = 1.0;
+        current_res = 1.0;
+        std::cout << "*****************************************" << std::endl;
+        std::cout << "************  refinement = " << refinement << " ************ " << std::endl;
+        std::cout << "viscosity= " << viscosity << std::endl;
+        std::cout << "*****************************************" << std::endl;
+
+        while ((first_step || (current_res > tolerance)) && outer_iteration < max_iteration)
+          {
+            if (first_step)
+              {
+                setup_dof();
+                initialize_system();
+                evaluation_point = present_solution;
+                assemble_NavierStokes_system(first_step, true, true);
+                solve(first_step);
+                present_solution = newton_update;
+                nonzero_constraints.distribute(present_solution);
+                first_step = false;
+                evaluation_point = present_solution;
+                assemble_NavierStokes_system(first_step, false, true);
+                current_res = system_rhs.l2_norm();
```

Comment:
That's clever how you solve the initial Stokes step by using a zero velocity field and different constraints. Learned something new :+1: